src/common/filter/FunctionManager.cpp

```diff
@@ -243,6 +243,156 @@ FunctionManager::FunctionManager() {
+        attr.minArity_ = 2;
+        attr.maxArity_ = 2;
+        attr.body_ = [] (const auto &args) {
+            auto value = Expression::asString(args[0]);
+            auto length = Expression::asInt(args[1]);
+            if (length < 0) {
+                length = 0;
+            }
+            return value.substr(0, length);
+        };
+    }
+    {
+        auto &attr = functions_["right"];
+        attr.minArity_ = 2;
+        attr.maxArity_ = 2;
+        attr.body_ = [] (const auto &args) {
+            auto value  = Expression::asString(args[0]);
+            auto length = Expression::asInt(args[1]);
+            if (length <= 0) {
+                length = 0;
+            }
+            return value.substr(value.size() - length);
+        };
+    }
+    {
+        auto &attr = functions_["lpad"];
+        attr.minArity_ = 3;
+        attr.maxArity_ = 3;
+        attr.body_ = [] (const auto &args) {
+            auto value = Expression::asString(args[0]);
+            size_t size  = Expression::asInt(args[1]);
+
+            if (size < 0) {
+                return std::string("");
+            } else if (size < value.size()) {
+                return value.substr(0, static_cast<int32_t>(size));
+            } else {
+                auto extra = Expression::asString(args[2]);
+                size -= value.size();
+                std::stringstream stream;
+                while (size > extra.size()) {
+                    stream << extra;
+                    size -= extra.size();
+                }
+                stream << extra.substr(0, size);
+                stream << value;
+                return stream.str();
+            }
+        };
+    }
+    {
+        auto &attr = functions_["rpad"];
+        attr.minArity_ = 3;
+        attr.maxArity_ = 3;
+        attr.body_ = [] (const auto &args) {
+            auto value = Expression::asString(args[0]);
+            size_t size  = Expression::asInt(args[1]);
+            if (size < 0) {
+                return std::string("");
+            } else if (size < value.size()) {
+                return value.substr(0, static_cast<int32_t>(size));
+            } else {
+                auto extra = Expression::asString(args[2]);
+                std::stringstream stream;
+                stream << value;
+                size -= value.size();
+                while (size > extra.size()) {
+                    stream << extra;
+                    size -= extra.size();
+                }
+                stream << extra.substr(0, size);
+                return stream.str();
+            }
+        };
+    }
+    {
+        auto &attr = functions_["substr"];
+        attr.minArity_ = 3;
+        attr.maxArity_ = 3;
+        attr.body_ = [] (const auto &args) {
+            auto value   = Expression::asString(args[0]);
+            auto start   = Expression::asInt(args[1]);
+            auto length  = Expression::asInt(args[2]);
+            return value.substr(start, length);
```

Comment:
Maybe need to handle exception if start > value.size()

Reply:
Good catch. 

According to `MySQL` , when the start is bigger than value's size or length is less than or equal to 0 . It will return a empty string.


```
mysql> select substr("123", 1, 1);
+---------------------+
| substr("123", 1, 1) |
+---------------------+
| 1                   |
+---------------------+
1 row in set (0.00 sec)

mysql> select substr("123", 1, 0);
+---------------------+
| substr("123", 1, 0) |
+---------------------+
|                     |
+---------------------+
1 row in set (0.00 sec)

mysql> select substr("123", 1, -1);
+----------------------+
| substr("123", 1, -1) |
+----------------------+
|                      |
+----------------------+
1 row in set (0.00 sec)

mysql> select substr("123", 5, 1);
+---------------------+
| substr("123", 5, 1) |
+---------------------+
|                     |
+---------------------+
1 row in set (0.00 sec)

mysql> select substr("123", -1, 1);
+----------------------+
| substr("123", -1, 1) |
+----------------------+
| 3                    |
+----------------------+
1 row in set (0.00 sec)

mysql> select substr("123", -1, 0);
+----------------------+
| substr("123", -1, 0) |
+----------------------+
|                      |
+----------------------+
1 row in set (0.00 sec)

mysql> select substr("123", -1, -1);
+-----------------------+
| substr("123", -1, -1) |
+-----------------------+
|                       |
+-----------------------+
1 row in set (0.00 sec)
```

When

Reply:
Reasonable.